In [2]:
import numpy as np

# Практика 1
Корешков

## Задача 1
Заокруглюючи наступні числа до трьох значущих цифр, 
визначити абсолютну та відносну похибки наближених чисел:


In [3]:
# 1)
a = 3.2523
# Залишаємо 3 значущі цифри. Останню збільшуємо на 1 якщо наступна за нею не менша 5
a1 = 3.25 # округлено

# Обчислення
d = a1 - a # абс похибка
e = d/a # відн похибка
print(f"абс похибка: {d:.3e}")
print(f"відн похибка: {e:.3e}")

абс похибка: -2.300e-03
відн похибка: -7.072e-04


## Задача 2
Визначити кількість вірних цифр в числі x, якщо його
відносна похибка


In [4]:
# 1)
x = 2.7981
e = 0.001 # відн похибка

# обчислення
d = x * e # абс похибка
print(f"абс похибка: {d}")
print(f"x-d ~ {x-d:.6f}; x+d ~ {x+d:.6f}")

абс похибка: 0.0027981
x-d ~ 2.795302; x+d ~ 2.800898


Абсолютна похибка $0.0028 < 0.005 = \frac{0.01}{2}$, отже вірним будуть розряди з нульового по $-2$.
Маємо 3 вірні цифри. Демонстрація:
 
$x-d \approx 2.80$  
$x+d \approx 2.80$  
але  
$x-d \approx 2.795$  
$x+d \approx 2.801$

## Задача 16.1
Оцінити абсолютну та відносну похибки обчислення
функції
$$f(x,y,z) = \ln \frac{xy}{z}$$
$$x=2.34, \quad y=1.25, \quad z=3.05$$
$$\Delta x = 0.01, \quad \Delta y = 0.02, \quad \Delta z = 0.02$$

Маємо пряму задачу. Спочатку обчислюємо часткові похідні $f$ за усіма змінними

$$f'_x(x,y,z) = \left(\frac{xy}{z}\right)^{-1} \cdot \frac{y}{z} = \frac{1}{x}$$
$$f'_y(x,y,z) = \left(\frac{xy}{z}\right)^{-1} \cdot \frac{x}{z} = \frac{1}{y}$$
$$f'_z(x,y,z) = \left(\frac{xy}{z}\right)^{-1} \cdot (-1)\cdot\frac{xy}{z^2} = -\frac{1}{z}$$

Далі припускаємо що похибки значення функції залежать від похибок аргументів наступним чином:
$$\Delta f(\vec x) = \sum_{i} f'_{x_i}(\vec x) \cdot \Delta x_i $$ 

І обчислюємо


In [5]:
x = np.array([2.34,1.25,3.05])
dx = np.array([0.01, 0.02, 0.02])
df = np.array([1/x[0], 1/x[1], -1/x[2]]).dot(dx)
f = np.log(x[0]*x[1]/x[2])
print(f"df = {df} ~ {df:.2f}")
print(f"f = {f}  ~ {f:.2f} +- {df:.2f}")

rdf = df/f
print(f"relative f error = {rdf:.5f}")

df = 0.013716127224323946 ~ 0.01
f = -0.04184710993550049  ~ -0.04 +- 0.01
relative f error = -0.32777


Зауважу, що я округлив абсолютну похибку функції до однієї значущої цифри, бо стільки цифр мали похибки в умові задачі. А далі округлив значення функції до кількості значущих цифр, що відповідає отриманій абсолютній похибці